File: DSA/Tree/Code/BinaryTree.cpp
#include <iostream>
#include <queue>
using namespace std;

// Binary Tree Node
struct Node {
    int data;
    Node* left;
    Node* right;
    
    Node(int val) : data(val), left(NULL), right(NULL) {}
};

class BinaryTree {
private:
    Node* root;
    
    Node* insertHelper(Node* node, int data) {
        if (node == NULL) {
            return new Node(data);
        }
        
        queue<Node*> q;
        q.push(node);
        
        while (!q.empty()) {
            Node* temp = q.front();
            q.pop();
            
            if (temp->left == NULL) {
                temp->left = new Node(data);
                break;
            } else {
                q.push(temp->left);
            }
            
            if (temp->right == NULL) {
                temp->right = new Node(data);
                break;
            } else {
                q.push(temp->right);
            }
        }
        
        return node;
    }
    
    void inorderHelper(Node* node) {
        if (node == NULL) return;
        
        inorderHelper(node->left);
        cout << node->data << " ";
        inorderHelper(node->right);
    }
    
    void preorderHelper(Node* node) {
        if (node == NULL) return;
        
        cout << node->data << " ";
        preorderHelper(node->left);
        preorderHelper(node->right);
    }
    
    void postorderHelper(Node* node) {
        if (node == NULL) return;
        
        postorderHelper(node->left);
        postorderHelper(node->right);
        cout << node->data << " ";
    }
    
    Node* searchHelper(Node* node, int data) {
        if (node == NULL) return NULL;
        
        if (node->data == data) return node;
        
        Node* left = searchHelper(node->left, data);
        if (left != NULL) return left;
        
        return searchHelper(node->right, data);
    }
    
public:
    BinaryTree() : root(NULL) {}
    
    void insert(int data) {
        root = insertHelper(root, data);
    }
    
    void inorder() {
        inorderHelper(root);
    }
    
    void preorder() {
        preorderHelper(root);
    }
    
    void postorder() {
        postorderHelper(root);
    }
    
    bool search(int data) {
        return searchHelper(root, data) != NULL;
    }
};

int main() {
    cout << "=== Binary Tree (C++) ===\n\n";
    
    BinaryTree bt;
    
    cout << "Inserting elements: 1, 2, 3, 4, 5, 6, 7\n";
    bt.insert(1);
    bt.insert(2);
    bt.insert(3);
    bt.insert(4);
    bt.insert(5);
    bt.insert(6);
    bt.insert(7);
    
    cout << "\nInorder Traversal (Left-Root-Right): ";
    bt.inorder();
    cout << "\n";
    
    cout << "Preorder Traversal (Root-Left-Right): ";
    bt.preorder();
    cout << "\n";
    
    cout << "Postorder Traversal (Left-Right-Root): ";
    bt.postorder();
    cout << "\n\n";
    
    cout << "Searching for 5: " << (bt.search(5) ? "Found" : "Not Found") << "\n";
    cout << "Searching for 10: " << (bt.search(10) ? "Found" : "Not Found") << "\n\n";
    
    cout << "=== Binary Tree Structure ===\n";
    cout << "Tree structure:\n";
    cout << "         1\n";
    cout << "        / \\\n";
    cout << "       2   3\n";
    cout << "      / \\ / \\\n";
    cout << "     4  5 6  7\n\n";
    
    cout << "=== Operations Complexity ===\n";
    cout << "Insert:     O(n) in worst case\n";
    cout << "Search:     O(n) in worst case\n";
    cout << "Traversal:  O(n) for all traversals\n";
    cout << "Space:      O(h) where h = height (recursion stack)\n";
    
    return 0;
}

File: DSA/Tree/Code/BinaryTree.py
# Binary Tree in Python

from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None
    
    def insert(self, data):
        """Insert element in level order"""
        if self.root is None:
            self.root = Node(data)
            return
        
        queue = deque([self.root])
        
        while queue:
            node = queue.popleft()
            
            if node.left is None:
                node.left = Node(data)
                break
            else:
                queue.append(node.left)
            
            if node.right is None:
                node.right = Node(data)
                break
            else:
                queue.append(node.right)
    
    def inorder(self, node):
        """Inorder traversal: Left-Root-Right"""
        if node is None:
            return
        
        self.inorder(node.left)
        print(node.data, end=" ")
        self.inorder(node.right)
    
    def preorder(self, node):
        """Preorder traversal: Root-Left-Right"""
        if node is None:
            return
        
        print(node.data, end=" ")
        self.preorder(node.left)
        self.preorder(node.right)
    
    def postorder(self, node):
        """Postorder traversal: Left-Right-Root"""
        if node is None:
            return
        
        self.postorder(node.left)
        self.postorder(node.right)
        print(node.data, end=" ")
    
    def search(self, node, data):
        """Search for element in binary tree"""
        if node is None:
            return False
        
        if node.data == data:
            return True
        
        left = self.search(node.left, data)
        if left:
            return True
        
        return self.search(node.right, data)

# Test
print("=== Binary Tree (Python) ===\n")

bt = BinaryTree()

print("Inserting elements: 1, 2, 3, 4, 5, 6, 7")
for i in range(1, 8):
    bt.insert(i)

print("\nInorder Traversal (Left-Root-Right): ", end="")
bt.inorder(bt.root)
print()

print("Preorder Traversal (Root-Left-Right): ", end="")
bt.preorder(bt.root)
print()

print("Postorder Traversal (Left-Right-Root): ", end="")
bt.postorder(bt.root)
print("\n")

print(f"Searching for 5: {'Found' if bt.search(bt.root, 5) else 'Not Found'}")
print(f"Searching for 10: {'Found' if bt.search(bt.root, 10) else 'Not Found'}\n")

print("=== Binary Tree Structure ===")
print("Tree structure:")
print("         1")
print("        / \\")
print("       2   3")
print("      / \\ / \\")
print("     4  5 6  7\n")

print("=== Operations Complexity ===")
print("Insert:     O(n) in worst case")
print("Search:     O(n) in worst case")
print("Traversal:  O(n) for all traversals")
print("Space:      O(h) where h = height (recursion stack)")
